src/openmp-utils.c

```diff
@@ -50,12 +51,20 @@ void initDTthreads() {
   ans = imin(ans, getIntEnv("OMP_THREAD_LIMIT", INT_MAX));  // user might expect `Sys.setenv(OMP_THREAD_LIMIT=2);setDTthreads()` to work. Satisfy this
   ans = imin(ans, getIntEnv("OMP_NUM_THREADS", INT_MAX));   //   expectation by reading them again now. OpenMP just reads them on startup (quite reasonably)
   DTthreads = ans;
+  DTthrottle = imax(1, getIntEnv("R_DATATABLE_THROTTLE", 1024)); // 2nd thread is used only when n>1024, 3rd thread when n>2048, etc
 }
 
-int getDTthreads() {
-  // this is the main getter used by all parallel regions; they specify num_threads(getDTthreads())
+int getDTthreads(const int64_t n, const bool throttle) {
+  // this is the main getter used by all parallel regions; they specify num_threads(n, DTthrottle) or num_threads(n, 1); 
   // Therefore keep it light, simple and robust. Local static variable. initDTthreads() ensures 1 <= DTthreads <= omp_get_num_proc()
-  return DTthreads;
+  // throttle is the number of iterations per thread before a second thread is utilized.
+  // throttle==1 when parallel region is already pre-chunked such as in fread; e.g. if there are 2 iterations it is intended for two
+  // threads to receive one iteration each. Otherwise, DTthrottle (by default 1024) is passed for standard loops over nrow (for example)
+  // where we leave OpenMP to batch. Passing the thottle directly avoids needing to know codes, and allows the calling parallel region
+  // to use a multiple of throttle (for example).
+  if (n<1) return 1;
+  int n2 = n>INT_MAX ? INT_MAX : n;
+  return imin(DTthreads, throttle ? 1+(n2-1)/DTthrottle : n2);
```

Comment:
It would help to document here in comment how the formula is meant to work. I don't have doubt it is the good one, but for anyone trying to understand why it is like this it may be challanging. I just had such problem understanding how batches in fsort/forder/gsum are defined.

Reply:
Ok will do.